### PR TITLE
feat: Only flycheck workspace that belongs to saved file

### PR DESCRIPTION
**Pull request by Veykril**
Supercedes https://github.com/rust-lang/rust-analyzer/pull/11038

There is still the problem that all the diagnostics are cleared, only clearing diagnostics of the relevant workspace isn't easily doable though I think, will have to dig into that

### CODE REVIEW

**Comment by Veykril** [comment]
Okay so I got a fix for the remaining problem, but somehow cargo is refusing to send diagnostics at some point. It works once and then never again for any but the last workspace...

**Comment by Veykril** [comment]
Ye so `cargo check` just finishes without errors for all but the last workspace when repeatedly invoked even though those workspaces contain errors 😕

File: crates/rust-analyzer/src/main_loop.rs [hunk 1]
```diff
@@ -720,13 +721,62 @@ impl GlobalState {
...
+                                        },
+                                    )
+                                }
+                                project_model::ProjectWorkspace::Json { project, .. } => project
+                                    .crates()
+                                    .any(|(c, _)| crate_ids.iter().any(|&crate_id| crate_id == c)),
+                                project_model::ProjectWorkspace::DetachedFiles { .. } => false,
+                            });
+                        'workspace: for (id, _) in workspace_ids {
+                            for flycheck in &this.flycheck {
```
Comment (bjorn3) [inline crates/rust-analyzer/src/main_loop.rs:761]: Maybe swap the loops? That is more cache friendly I think and avoids the need for `continue 'workspace;`.

**Comment by bjorn3** [comment]
What if one workspace depends on a crate from the other workspace? For example if I edit cranelift-codegen it would be nice if cg_clif gets recompiled. I can't put cranelift-codegen in cg_clif's workspace as it is already in a different workspace and as it is used as `[patch.crates-io]` replacement rather than a regular `path = "..."` dependency.

File: crates/rust-analyzer/src/main_loop.rs [hunk 2]
```diff
@@ -720,13 +721,62 @@ impl GlobalState {
...
+                                    })
+                                    .transpose()
+                            })
+                            .collect::<ide::Cancellable<_>>()?;
+                        let paths: Vec<_> = paths.iter().map(Deref::deref).collect();
+
+                        let workspace_ids =
+                            this.workspaces.iter().enumerate().filter(|(_, ws)| match ws {
+                                project_model::ProjectWorkspace::Cargo { cargo, .. } => {
+                                    cargo.packages().filter(|&pkg| cargo[pkg].is_member).any(
```
Comment (bjorn3) [inline crates/rust-analyzer/src/main_loop.rs:747]: I think removing the `is_member` filter here would fix the above scenario.

**Comment by Veykril** [comment]
Okay, so I think I am hitting a cargo bug now? I basically have three workspaces all of which have errors in them. When r-a starts up it starts a flycheck process for each and we get diagnostics for all three. But due to how r-a works we redo this a few times at the start (like 2 or 3 times due to VFS change reloads), now here is where it gets weird, all following flychecks report successfull builds on the first two workspaces while the last one still errors, and this keeps happening unless the workspaces are flychecked on their own. So it seems cargo gets confused when multiple checks are running at the same time?

**Comment by lnicola** [comment]
Why, yes. I actually complained about this, thinking it's a RA bug:

image

**Comment by Veykril** [comment]
Lovely, so I guess we can't do this until that gets fixed. I assume there is no issue on the cargo issue tracker regarding this yet?

**Comment by lnicola** [comment]
The thing is, I can't reproduce this (outside of RA) with `cargo check --all-targets; cargo check`.

**Comment by lnicola** [comment]
We also have the `RUSTC_WRAPPER` thing, but I can't disable it ("failed to run build scripts").

**Comment by Veykril** [comment]
Turns out triggering the cargo bug doesnt *always* occur and I changed some things around so we don't start a bunch of flycheck in the middle of fetching the workspace which should surpress the bug from occuring mostly. I'll merge this for now, we'll see if this causes trouble though from my testing it should work fine.
@bors r+

File: crates/rust-analyzer/src/global_state.rs [hunk 3]
```diff
@@ -192,6 +192,7 @@ impl GlobalState {
                 if let Some(path) = vfs.file_path(file.file_id).as_path() {
                     let path = path.to_path_buf();
                     if reload::should_refresh_for_change(&path, file.change_kind) {
+                        tracing::warn!("fetch-fiel_change");
```
Comment (lnicola) [inline crates/rust-analyzer/src/global_state.rs:195]: Should be removed or changed to `info!`?

Comment (Veykril) [reply crates/rust-analyzer/src/global_state.rs:195]: was already removed in a follow up

Comment (lnicola) [reply crates/rust-analyzer/src/global_state.rs:195]: Yeah, I actually missed it in #12947

**Pull request merged**